File: cpp/trainKB.cpp
#include <iostream>
#include <string>
#include <vector>
#include <atomic>
#include <thread>
#include <chrono>
#include <unordered_map>
#include <cmath>
#include <utility>

#include "optparse.h"
#include "ReaderLines.h"
#include "RandomGenerator.h"
#include "Poisson.h"
#include "TrainerKB.h"
#include "MultinomialTable.h"
#include "misc.h"

using namespace std;
using namespace misc;

class option : public optparse {
public:
  bool help = false;

  double sampPow = 0.75;
  double sampPathLen = 0.5;
  long long numBatches = 1000000;
  const char* inPath = nullptr;
  string outPath;
  int para = 2;

  BEGIN_OPTION_MAP()
    ON_OPTION(SHORTOPT('h') || LONGOPT("help"))
      help = true;
    ON_OPTION_WITH_ARG(LONGOPT("sampPow"))
      sampPow = stod(arg);
    ON_OPTION_WITH_ARG(LONGOPT("sampPathLen"))
      sampPathLen = stod(arg);
    ON_OPTION_WITH_ARG(LONGOPT("numBatches"))
      numBatches = stoll(arg);
    ON_OPTION_WITH_ARG(LONGOPT("inPath"))
      inPath = arg;
    ON_OPTION_WITH_ARG(LONGOPT("outPath"))
      outPath = string(arg);
    ON_OPTION_WITH_ARG(LONGOPT("para"))
      para = stoi(arg);

  END_OPTION_MAP()
};

static vector<vector<pair<unsigned int, unsigned int>>> graph; // neighbors: (relation_index, tail_index)
static MultinomialTable samp_node;
static atomic_ullong remained_batches;

static void trainKB_para(int tid, RandomGenerator rnd, double pl, TrainerKB* ptrain) {
  Poisson samp_path(pl);

  long long remained;
  while((remained = remained_batches.fetch_sub(1, memory_order_relaxed)) > 0) {
    if (remained % 100000 == 0) {
      cerr << remained << endl;
    }
    unsigned int hi = samp_node.sample(rnd);

    vector<vector<pair<unsigned int, unsigned int>>> pths;
    unsigned int samp_sz = 0;
    const auto& neighbor = graph[hi];
    for (unsigned int i = 0; i != neighbor.size() * 2; ++i) {
      vector<pair<unsigned int, unsigned int>> pth;
      auto edge = neighbor[rnd(neighbor.size())];
      samp_path.reset();
      do {
        pth.push_back(edge);
        if (++samp_sz == 31) break;
        const auto& nei = graph[edge.second];
        edge = nei[rnd(nei.size())];
      } while (!samp_path.stop(rnd));
      pths.push_back(move(pth));
      if (samp_sz == 31) break;
    }
    ptrain->update(rnd, hi, pths);
  }
}

int main(int argc, char *argv[])
{
  try {
    option opt;
    int argpos = opt.parse(argv, argc);
    if (opt.help) {
      cout << "Train model for KB." << endl
           << "  trainKB [OPTION...] VOCAB_ENTITY VOCAB_RELATION TRAIN_FILE" << endl
           << endl << "positional arguments:" << endl
           << "  VOCAB_ENTITY      counts of entities" << endl
           << "  VOCAB_RELATION    counts of relations" << endl
           << "  TRAIN_FILE        train file" << endl
           << endl << "optional arguments:" << endl
           << "  -h, --help        show this help message and exit" << endl
           << "  --sampPow         samp. node prob. is power of freq. (default: 0.75)" << endl
           << "  --sampPathLen     path length is 1+Poisson(sampPathLen) (default: 0.5)" << endl
           << "  --numBatches      batches to train (default: 1000000)" << endl
           << "  --inPath          if set, load model from this path for init" << endl
           << "  --outPath         save model to this path (default: working dir)" << endl
           << "  --para            number of parallel threads (default: 2)" << endl
          ;
      return 0;
    }
    if (argc - argpos != 3) throw runtime_error("wrong number of arguments");
    string words_fn(argv[argpos]);
    string roles_fn(argv[argpos + 1]);
    string train_fn(argv[argpos + 2]);

    // read vocab of entities
    unordered_map<string, unsigned int> words;
    unsigned int wsz = 0; {
      vector<double> wprobs;
      ReaderLines wlines(words_fn);
      while (!wlines.empty()) {
        auto sp = split(wlines.next(), '\t');
        words[sp[0]] = wsz;
        ++wsz;
        wprobs.push_back(stod(sp[1]));
      }
      for (auto& x : wprobs) x = pow(x, opt.sampPow);
      samp_node = MultinomialTable(wprobs.cbegin(), wprobs.cend(), 1 << 16);
    }

    // read vocab of relations
    unordered_map<string, unsigned int> roles;
    unsigned int rsz = 0; {
      ReaderLines rlines(roles_fn);
      while (!rlines.empty()) {
        roles[split(rlines.next(), '\t')[0]] = rsz;
        ++rsz;
      }
    }

    //read train file, add neighbors to graph
    graph.resize(wsz);
    ReaderLines glines(train_fn);
    while (!glines.empty()) {
      auto sp = split(glines.next(), '\t');
      const unsigned int head_index = words.at(sp[0]);
      const unsigned int tail_index = words.at(sp[2]);
      const unsigned int rel_index = roles.at(sp[1]);
      graph[head_index].emplace_back(rel_index, tail_index);
      graph[tail_index].emplace_back(rel_index + rsz, head_index);
    }

    RandomGenerator rg(static_cast<uint64_t>(chrono::system_clock::now().time_since_epoch().count()));

    TrainerKB trainer;
    trainer.saveParams(opt.outPath);
    if (opt.inPath) trainer.loadModel(wsz, rsz, opt.inPath);
    else {
      trainer.initModel(wsz, rsz, rg);
      trainer.saveModel(opt.outPath + "init_");
    }

    vector<thread> threads;
    threads.reserve(opt.para);

    remained_batches = opt.numBatches;
    for (int i = 0; i != opt.para; ++i) {
      rg.jump();
      threads.emplace_back(&trainKB_para, i, rg, opt.sampPathLen, &trainer);
    }
    for (auto& x : threads) x.join();

    trainer.saveModel(opt.outPath);

  } catch (const optparse::unrecognized_option& e) {
    cout << "unrecognized option: " << e.what() << endl;
    return 1;
  } catch (const optparse::invalid_value& e) {
    cout << "invalid value: " << e.what() << endl;
    return 1;
  } catch (const exception& e) {
    cout << "use -h or --help to show help." << endl;
    cout << e.what() << endl;
  }

  return 0;
}

File: python/trainKB.py
# -*- coding: utf-8 -*-

import argparse
import math
import numpy as np
import importlib
import importlib.util

from utilityFuncs import readerLine

def main():
  parser = argparse.ArgumentParser(description='Train model for KB.')
  parser.add_argument('words_file', metavar='VOCAB_ENTITY', type=str,
                      help='counts of entities')
  parser.add_argument('roles_file', metavar='VOCAB_RELATION', type=str,
                      help='counts of relations')
  parser.add_argument('train_file', metavar='TRAIN_FILE', type=str,
                      help='train file')

  parser.add_argument('--sampPow', dest='sampPow', type=float, default=0.75,
                      help='sampling nodes by probabilities proportional to the power of frequency. (default: 0.75)')
  parser.add_argument('--sampPathLen', dest='sampPathLen', type=float, default=0.5,
                      help='path length is 1+Poisson(sampPathLen) (default: 0.5)')
  parser.add_argument('--numBatches', dest='numBatches', type=int, default=1000000,
                      help='batches to train (default: 1000000)')
  parser.add_argument('--inPath', dest='inPath', type=str, default=None,
                      help='if set, load model from this path for init')
  parser.add_argument('--outPath', dest='outPath', type=str, default="",
                      help='save model to this path (default: working dir)')
  parser.add_argument('--para', dest='para', type=int, default=2,
                      help='number of parallel threads (default: 2)')
  parser.add_argument('--glimvecModule', dest='glimvecModule', type=str, default=None,
                      help='path to the pre-trained python library (default: None)')

  args = parser.parse_args()

  # read vocab of entities
  words = {}
  wsz = 0
  wprobs = []
  for line in readerLine(args.words_file):
    w, freq = line.split('\t')
    words[w] = wsz
    wsz += 1
    wprobs.append(math.pow(float(freq), args.sampPow))
  samp_node_prob = np.array(wprobs) / np.sum(wprobs)

  # read vocab of relations
  roles = {}
  rsz = 0
  for line in readerLine(args.roles_file):
    roles[line.split('\t')[0]] = rsz
    rsz += 1

  # read train file, add neighbors to graph
  graph = [[]] * wsz
  for line in readerLine(args.train_file):
    head, rel, tail = line.split('\t')
    head_index = words[head]
    rel_index = roles[rel]
    tail_index = words[tail]
    graph[head_index].append((rel_index, tail_index))
    graph[tail_index].append((rel_index + rsz, head_index))

  # function for generating batch
  #  sample size in a batch should not exceed 31
  def genBatch(tid):
    hi = np.random.choice(wsz, p=samp_node_prob)
    pths = []
    samp_sz = 0
    neighbor = graph[hi]
    for i in range(len(neighbor) * 2):
      pth = []
      edge = neighbor[np.random.choice(len(neighbor))]
      pth.append(edge)
      samp_sz += 1
      if samp_sz < 31:
        for _ in range(np.random.poisson(args.sampPathLen)):
          nei = graph[edge[1]]
          edge = nei[np.random.choice(len(nei))]
          pth.append(edge)
          samp_sz += 1
          if samp_sz == 31:
            break
      pths.append(pth)
      if samp_sz == 31:
        break
    return hi, pths

  # to debug, first check if the genBatch function is ok:
  #print(genBatch(0))

  if args.glimvecModule is None:
    glimvec = importlib.import_module('glimvec')
  else:
    spec = importlib.util.spec_from_file_location("glimvec", args.glimvecModule)
    glimvec = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(glimvec)

  glimvec.initTrainer(wsz, rsz, inPath=args.inPath, outPath=args.outPath)
  glimvec.trainKB(genBatch, args.numBatches, args.para)
  glimvec.saveModel(args.outPath)


if __name__ == '__main__':
  main()
